### PR TITLE
feat(optimizer): Add output indices to `LogicalJoin`

### PR DESCRIPTION
**Pull request by jon-chuang**
## What's changed and what's your intention?
Adds output indices to `LogicalJoin` to avoid simple projections which merely reorder columns.

Future:
- Add more optimizer rules to merge simple projections into `LogicalJoin`s/other operators with `output_indices`.
- Impl output_indices in plan proto and executors. For streaming hash join, we can just modify the chunk builder to take in an `output_indices` parameter.
- `LogicalMultiJoin` also has `output_indices`. It needs to merge output indices and rewrite conditions. 

## Refer to a related PR or issue link (optional)
Tracked in: https://github.com/singularity-data/risingwave/issues/1922

### CODE REVIEW

**Review by Enter-tainer — APPROVED** [review]
Generally LGTM.

BTW, just to confirm, for o2l l2o r2o o2r... mappings, the "o" here should be "original" rather than "output", right? Maybe the comments should be tweaked. 

cc @st1page PTAL.

**Comment by jon-chuang** [comment]
> BTW, just to confirm, for o2l l2o r2o o2r... mappings, the "o" here should be "original" rather than "output", right? Maybe the comments should be tweaked.

Yes, I think I try to maintain the concept of `internal_column_number`. Perhaps it should be renamed to `i2l`, `l2i`, and we also introduce `i2o` which is the mapping defined by the `output_indices`. I used this concept of "internal" columns in certain parts of this PR. I think I'll adopt it and add some comments.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 1]
```diff
@@ -621,24 +726,74 @@ impl ToStream for LogicalJoin {
...
+            .filter(|i| l2o.try_map(*i) == None);
+
+        let right_to_add = right
+            .pk_indices()
+            .iter()
+            .cloned()
+            .filter(|i| r2o.try_map(*i) == None)
+            .map(|i| i + left_len);
+
+        let mut new_output_indices = join.output_indices.clone();
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:786]: Add comments that we are adding PK indices to the output_indices

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 2]
```diff
@@ -44,28 +44,44 @@ pub struct LogicalJoin {
...
 }
 
 impl LogicalJoin {
     pub(crate) fn new(left: PlanRef, right: PlanRef, join_type: JoinType, on: Condition) -> Self {
+        let out_column_num =
+            Self::out_column_num(left.schema().len(), right.schema().len(), join_type);
+        Self::new_with_reordered_output(left, right, join_type, on, (0..out_column_num).collect())
+    }
+
+    pub(crate) fn new_with_reordered_output(
```
Comment (Enter-tainer) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:69]: `output_indices` can not only handle reorder but also column removals. Maybe `new_with_output_indices`?

**Comment by st1page** [comment]
>Yes, I think I try to maintain the concept of internal_column_number. Perhaps it should be renamed to i2l, l2i, and we also introduce i2o which is the mapping defined by the output_indices. I used this concept of "internal" columns in certain parts of this PR. I think I'll adopt it and add some comments.

could you please do it in this PR? I am afraid that some others use the "o2l, l2o" such as distribution derive or some rules. so we might be opposed to guaranteed current semantics.

**Comment by jon-chuang** [comment]
Hi! ready for final review. I applied suggestions from the comments and also added some changes for affected components.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 3]
```diff
@@ -552,8 +657,12 @@ impl ToBatch for LogicalJoin {
         let left = self.left().to_batch()?;
         let right = self.right().to_batch()?;
         let logical_join = self.clone_with_left_right(left, right);
+        let new_output_indices = logical_join.output_indices.clone();
+        let new_internal_column_num = logical_join.internal_column_num();
+        let default_indices = (0..new_internal_column_num).collect::<Vec<_>>();
+        let logical_join = logical_join.clone_with_output_indices(default_indices.clone());
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:663]: Nice work! I did not think about that.

**Review by st1page — APPROVED** [review]
LGTM

File: src/frontend/test_runner/tests/testdata/join.yaml [hunk 4]
```diff
@@ -8,39 +8,42 @@
   logical_plan: |
     LogicalProject { exprs: [$1, $2, $4, $5, $7, $8] }
       LogicalFilter { predicate: ($1 = $4) AND ($1 = $7) }
-        LogicalJoin { type: Inner, on: true }
-          LogicalJoin { type: Inner, on: true }
+        LogicalJoin { type: Inner, on: true, output_indices: [0, 1, 2, 3, 4, 5, 6, 7, 8] }
```
Comment (skyzh) [inline src/frontend/test_runner/tests/testdata/join.yaml:11]: I think this should only be shown when it isn't `0, 1, 2, ...`

**Comment by skyzh** [comment]
... and there's some problems with the plan in TPC-H:

```
                        StreamProject { exprs: [$1, $0, $2, $3, $4, $5] }
                           StreamProject { exprs: [$0, $1, $3, $4, $5, $7] }
```

It seems that two projections are not combined? Is it because of rule application order?

**Comment by Enter-tainer** [comment]
> It seems that two projections are not combined? Is it because of rule application order?

A workaround is discussed in https://github.com/singularity-data/risingwave/pull/2769#discussion_r883316583 . And I'm working on pushing output indices to executors so the workaround will not last too long.

File: src/frontend/test_runner/tests/testdata/join.yaml (hunk 4 above)
Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/join.yaml:11]: makes sense!

**Comment by jon-chuang** [comment]
Yes, basically its due to a workaround!

**Comment by skyzh** [comment]
So we should merge https://github.com/singularity-data/risingwave/pull/2769 before merging this PR? 🤪

**Comment by Enter-tainer** [comment]
I think the order of merging is not a big issue because extra batch/stream project nodes don't hurt correctness :rofl: . And I agree with you that it would be better if we merge #2769 before this pr. BTW, works in #2769 are mostly done and hopefully can be merged soon. But if this PR is ready, I think it is OK to merge it.

**Comment by jon-chuang** [comment]
should be ready, I'll merge when the CI passes!

**Pull request merged**